### PR TITLE
#1071 `Commands` implemented + Language linked responses.

### PR DESCRIPTION
**Pull request by criske**
This PR also introduces linked language responses. This means that a value for a response key could be a link
to a separate file that contains the actual reply. The file, depending on used link protocol, could be fetched locally (`file://`, `classpath:`) or remotely (`http(s)://`).

By doing this, larger and/or marked up replies can be easly maintainted in separate files rather than having them clobbered in the `responses.properites`.

PR for #1071

### CODE REVIEW

**Comment by zoeself** [comment]
@criske thank you for your Pull Request. I'll assign someone to review it soon.

File: self-core-impl/src/test/java/com/selfxdsd/core/projects/LanguageLinkedResponsesTestCase.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+     * @checkstyle JavadocMethod (40 lines).
+     */
+    @Test
+    public void shouldReadReplyFromHttpLink() {
+        final Properties commands = new Properties();
+        commands.putIfAbsent(
+            "commands.comment",
+            "https://raw.githubusercontent.com/self-xdsd/self-docs/"
+                + "master/projectmanager.md"
+        );
```
Comment (criske) [inline self-core-impl/src/test/java/com/selfxdsd/core/projects/LanguageLinkedResponsesTestCase.java:101]: @amihaiemil It will be nice is we could extract the Commands part into a separate file and then having in the responses.properties
commands.comment = https://raw.githubusercontent.com/self-xdsd/self-docs/master/commands.md
This way we will have a centralized `commands` documentation used by both self-docs and self-pms.

**Comment by zoeself** [comment]
@amihaiemil I couldn't find any assignee for this task. This is either because there are no contributors with role ``REV`` available or because the project does not have enough funds.

Please, make sure there is at least one available contributor with the required role and the project can afford to pay them.

File: self-api/src/main/java/com/selfxdsd/api/Language.java [hunk 2]
```diff
@@ -56,6 +65,7 @@ public Language(
         final String commandsFileName,
         final String responsesFileName
     ) {
+        this(new Properties(), new Properties());
```
Comment (amihaiemil) [inline self-api/src/main/java/com/selfxdsd/api/Language.java:68]: @criske Why is this needed? :D 

I would prefer to leave the constructor as it was initially.

File: self-api/src/main/java/com/selfxdsd/api/Language.java [hunk 3]
```diff
@@ -104,7 +128,95 @@ public final String categorize(final String command) {
...
+     *
+     * If something goes wrong like IO exception, bad request etc...,
+     * it will return null.
+     *
+     * @param linkedReply Reply as link.
+     * @return Actual reply fetched from link or null if something goes wrong.
+     * @checkstyle CyclomaticComplexity (60 lines).
+     * @checkstyle BooleanExpressionComplexity (20 lines).
+     */
+    private String tryReplyFromLink(final String linkedReply){
```
Comment (amihaiemil) [inline self-api/src/main/java/com/selfxdsd/api/Language.java:173]: @criske Let's name this method ``followPossibleLink``. 

Also, I like the idea of linked files for bigger responses but this method is waaay over-engineered. The ``classpath://`` option is enough, please remove the rest of the logic. 

If we will ever read language resources from a webservice, we will need some serious architectural changes.

File: self-api/src/main/java/com/selfxdsd/api/Language.java [hunk 4]
```diff
@@ -104,7 +128,95 @@ public final String categorize(final String command) {
      * @return String reply or null if nothing is found.
      */
     public final String reply(final String key) {
-        return this.responses.getProperty(key);
+        final String reply = this.responses.getProperty(key);
+        return tryReplyFromLink(reply);
```
Comment (amihaiemil) [inline self-api/src/main/java/com/selfxdsd/api/Language.java:132]: @criske just ``return followPossibleLink(this.responses.getProperty(key));`` should be enough.

File: self-core-impl/src/test/resources/replies/commands.comment_en.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+* ``Hello`` -- any comment addressed to the PM containing the word **"hello"**.
+* ``Status`` -- any comment addressed to the PM, containing the word **"status"**. The PM will respond with the Task's status, assignee, deadline etc.
+* ``Resign`` -- a task's assignee can always resign from the task, if they cannot or simply do not want to solve it. The comment should be addressed to the PM and contain the word **"resign"**. Only the task's assignee can give this command.
+* ``Deregister`` -- a task can be deregistered (taken out of scope) with a comment containing the word **"deregister"**. Only a Contributor with role ``PO`` or ``ARCH`` can give this command. When a task is deregistered, the task's assignee is automatically resigned.
+* ``Register`` -- an Issue or PR can be registered as a Task with a comment containing the word **"register"**. Any Contributor can give this command.
```
Comment (amihaiemil) [inline self-core-impl/src/test/resources/replies/commands.comment_en.md:5]: @criske Can you also start this message with: ``Hi @%s! Here are the commands which I understand: ``, where %s should be replaced with the author's username. 

Also, at Register and Deegister, we should specify ``add`` and ``remove`` too. If I remember well, also the Resign command has more options.

If you look into ``commands_en.properties``, some commands have more than one entry, which means the PM will respond with that command for all those entries.

**Review by amihaiemil — CHANGES_REQUESTED** [review]
@criske looks nice overall, but I have some comments :D

**Comment by amihaiemil** [comment]
@criske one more thing: I don't see the ``Commands`` class being used anywhere. Did you forget to add it in the Conversation composition? :D

**Comment by amihaiemil** [comment]
@criske inca o chestie, mi-am amintit acum. Fisierele astea de proprietati si raspunsuri in core sunt resurse de Test. Ele trebuie copiate in self-pm ca sa functioneze. Deci mecanismu asta cu classpath:// ar trebui sa functioneze si acolo :D

File: self-api/src/main/java/com/selfxdsd/api/Language.java (hunk 3 above)
Comment (criske) [reply self-api/src/main/java/com/selfxdsd/api/Language.java:173]: >The classpath:// option is enough, please remove the rest of the logic.

@amihaiemil  I agree the `file://` should be removed, classpath is enough for local files.

> If we will ever read language resources from a webservice, we will need some serious architectural changes.

@amihaiemil For this particular case, https://github.com/self-xdsd/self-docs is the webservice.
We only need to extract "Commands" listing part from https://raw.githubusercontent.com/self-xdsd/self-docs/master/projectmanager.md into a separate file, `commands.md` for ex. I think Jekyll supports including other files.
(something like `{%commands.md%}` in `projectmanager.md`)

Now on `self-pm` responses.properties we could have two entries.
```properties
commands.comment.header=Hi @%s! Here are the commands which I understand: 
commands.comment.body=https://raw.githubusercontent.com/self-xdsd/self-docs/master/commands.md
```
And then in `Commands` Conversation we merge the header with body to `SendReply`.
This way any changes on self-docs' `commands.md` will be reflected on the `commands` reply.

But until we have that refactoring on self-docs we could still use the `classpath` aproach. What do you think ? :D

File: self-api/src/main/java/com/selfxdsd/api/Language.java (hunk 2 above)
Comment (criske) [reply self-api/src/main/java/com/selfxdsd/api/Language.java:68]: @amihaiemil Needed to test `file://` protocol, that works only with absolute paths (I couldn't make it to work otherwise).
Just adding a file full path in `responses.properties` will break the ci though.

**Comment by criske** [comment]
> @criske one more thing: I don't see the `Commands` class being used anywhere. Did you forget to add it in the Conversation composition? :D

I forgot about that, sorry :D

File: self-api/src/main/java/com/selfxdsd/api/Language.java (hunk 3 above)
Comment (amihaiemil) [reply self-api/src/main/java/com/selfxdsd/api/Language.java:173]: @criske I see your point but I don't agree. Self-Docs is just a static site for documentation, I don't want to mix them up together. :D 

Someone will surely end up modifying self-docs without knowing of this mechanism and break it. Please only use the classpath option.

Comment (amihaiemil) [reply self-api/src/main/java/com/selfxdsd/api/Language.java:173]: @criske If you want to make an integration and make sure everything is up to date, then we should generate the documentation from the code, not the other way around :D 

Maybe we can think about this later

**Comment by amihaiemil** [comment]
@criske nice, thanks!

**Comment by amihaiemil** [comment]
@rultor merge
